haystack/components/preprocessors/utils.py

```diff
@@ -0,0 +1,228 @@
+        """
+        sentence_spans = list(self.sentence_tokenizer.span_tokenize(text))
+        if self.use_split_rules:
+            sentence_spans = self._apply_split_rules(text, sentence_spans)
+
+        sentences = [{"sentence": text[start:end], "start": start, "end": end} for start, end in sentence_spans]
+        return sentences
+
+    def _apply_split_rules(self, text: str, sentence_spans: List[Tuple[int, int]]) -> List[Tuple[int, int]]:
+        """
+        Applies additional split rules to the sentence spans.
+
+        :param text: The text to split.
+        :param sentence_spans: The list of sentence spans to split.
+        :returns: The list of sentence spans after applying the split rules.
+        """
+        new_sentence_spans = []
+        quote_spans = [match.span() for match in re.finditer(r"\W(\"+|\'+).*?\1", text)]
+        while sentence_spans:
+            span = sentence_spans.pop(0)
+            next_span = sentence_spans[0] if len(sentence_spans) > 0 else None
+            while next_span and self._needs_join(text, span, next_span, quote_spans):
+                sentence_spans.pop(0)
+                span = (span[0], next_span[1])
+                next_span = sentence_spans[0] if len(sentence_spans) > 0 else None
+            start, end = span
+            new_sentence_spans.append((start, end))
+        return new_sentence_spans
+
+    def _needs_join(
+        self, text: str, span: Tuple[int, int], next_span: Tuple[int, int], quote_spans: List[Tuple[int, int]]
+    ) -> bool:
+        """
+        Checks if the spans need to be joined as parts of one sentence.
+
+        :param text: The text containing the spans.
+        :param span: The current sentence span within text.
+        :param next_span: The next sentence span within text.
+        :param quote_spans: All quoted spans within text.
+        :returns: True if the spans needs to be joined.
+        """
+        start, end = span
+        next_start, next_end = next_span
+
+        # sentence. sentence"\nsentence -> no split (end << quote_end)
+        # sentence.", sentence -> no split (end < quote_end)
+        # sentence?", sentence -> no split (end < quote_end)
+        if any(quote_start < end < quote_end for quote_start, quote_end in quote_spans):
+            # sentence boundary is inside a quote
+            return True
+
+        # sentence." sentence -> split (end == quote_end)
+        # sentence?" sentence -> no split (end == quote_end)
+        if any(quote_start < end == quote_end and text[quote_end - 2] == "?" for quote_start, quote_end in quote_spans):
+            # question is cited
+            return True
+
+        if re.search(r"(^|\n)\s*\d{1,2}\.$", text[start:end]) is not None:
+            # sentence ends with a numeration
+            return True
+
+        # next sentence starts with a bracket or we return False
+        return re.search(r"^\s*[\(\[]", text[next_start:next_end]) is not None
```

Comment:
All of these linters and checkers didn't allow me to keep this LOC as it was in the original code. Please double check @sjrl

Reply:
What was the original line of code?

Reply:
It was:
```
if re.search(r"^\s*[\(\[]", text[next_start:next_end]) is not None:
    return True
    
return False
```

But then linter didn't allow me to have such code :-)